### PR TITLE
make it use php-fpm

### PR DESCRIPTION
**Pull request by jospoortvliet**
So I thought enabling fpm wouldn't be so hard ;-)

What a deep hole I found myself in... Anybody around here willing to dig in with me?

The basics are there but I haven't yet tested it. what is missing, for sure, is the init part - I'm not sure how to get an init script working.

What does this do?
- compile php-fpm in (in the snapcraft)
- make sure it starts as daemon (in the snapcraft)
- configure php-fpm (adding a config file)
- enable using php-fpm in apache (in the apache_config file)

I have not yet tested it so the chances this work are around freezing point. I did get it to work on my own server but that doesn't run snappy atm.

@kyrofa @oparoz @ezraholm @enoch85 what do you think, could any of you give this a try perhaps?

### CODE REVIEW

File: src/owncloud/apache_config [hunk 1]
```diff
@@ -14,3 +14,16 @@ Alias "/apps" "${SNAP_DATA}/owncloud/apps"
 <IfModule dir_module>
     DirectoryIndex index.html index.php
 </IfModule>
+
+<IfModule proxy_fcgi_module>
+	<Proxy "unix:/var/run/php.socket|fcgi://php-fpm" timeout=300>
+	</Proxy>
+</IfModule>
+
+<Directory "${SNAP}">
```
Comment (jospoortvliet) [inline src/owncloud/apache_config:24]: not sure what folder this should be tbh

File: src/owncloud/apache_config [hunk 2]
```diff
@@ -14,3 +14,21 @@ Alias "/apps" "${SNAP_DATA}/owncloud/apps"
...
+	<IfModule proxy_fcgi_module>
+		<FilesMatch \.php$>
+			SetHandler "proxy:fcgi://php-fpm/"
+		</FilesMatch>
+	</IfModule>
+</Directory>
+
+# this should route all php files though fpm - note that fpm has to run with the same rights as apache (to own the files)
+ProxyPassMatch ^/(.*\.php(/.*)?)$ fcgi://127.0.0.1:9000/srv/www/htdocs/owncloud/$1
+DirectoryIndex /index.php index.php
```
Comment (jospoortvliet) [inline src/owncloud/apache_config:34]: ... but I'm not entirely sure this has to be here. I know it has to end up in the vhost file of ownCloud (at least that's where I put it on my system) but is this going there?

**Comment by oparoz** [comment]
Cool, we're going to be able to compare approaches :d 

The info which I can't find is how to create users... Somehow Snappy has a www-data user and a bunch of ones we don't use, but the owncloud snap doesn't manage that at all.

**Comment by kyrofa** [comment]
> The info which I can't find is how to create users... Somehow Snappy has
a www-data user and a bunch of ones we don't use, but the owncloud snap
doesn't manage that at all.

No need to worry about users-- all Snappy services run as root. In typical
Linux installations this would obviously be dangerous, but in the case of
Snappy the services are very confined.

**Comment by kyrofa** [comment]
> The info which I can't find is how to create users... Somehow Snappy has a www-data user and a bunch of ones we don't use, but the owncloud snap doesn't manage that at all.

No need to worry about users-- all Snappy services run as root. In typical Linux installations this would obviously be dangerous, but in the case of Snappy the services are very confined.

**Comment by oparoz** [comment]
> No need to worry about users-- all Snappy services run as root. In typical
> Linux installations this would obviously be dangerous, but in the case of
> Snappy the services are very confined.

So a PHP application is root and can wipe the entire writeable folder?

**Comment by jospoortvliet** [comment]
btw docu i used are the php.net docs... http://php.net/manual/en/install.fpm.php

**Comment by oparoz** [comment]
@kyrofa - ~~Or just read ownCloud's config and have full access to the database~~ It does that already obviously :D

**Comment by kyrofa** [comment]
> So a PHP application is root and can wipe the entire writeable folder?

Only its own `$SNAP_DATA` folder-- no one else's. "running as root" in Snappy does not mean "running with absolute privilege." It's still confined.

**Comment by jospoortvliet** [comment]
@kyrofa so I should change php-fpm to have the root:root rights? And the data all has root:root too?

**Comment by kyrofa** [comment]
> ~~Or just read ownCloud's config and have full access to the database~~ It does that already obviously :D 

Right, exactly. Let's do a high-level comparision of a traditional ownCloud install (with [recommended permissions](https://doc.owncloud.org/server/9.0/admin_manual/installation/installation_wizard.html#strong-perms-label)) to a Snappy ownCloud install, and what a compromise might look like on the areas we care about: the ownCloud application, the ownCloud apps, the ownCloud config, the ownCloud data, and the ownCloud database.

#### Traditional

If ownCloud (or Apache) is compromised, the attacker gains the privileges of the process compromised, in this case probably `www-data`. The effect on our areas of concern:
- ownCloud application: read-access
- ownCloud apps: read-write access
- ownCloud config: read-write access
- ownCloud data: read-write access
- ownCloud database: read-write access (only to databases ownCloud's database user can access, however).

#### Snappy

Again, if ownCloud (or Apache) is compromised, the attacker gains the privileges of the process compromised, which in this case is _running_ as `root`, but is still confined to the snap's playground. The effect on our areas of concern:
- ownCloud application: read-access
- ownCloud apps: read-write access
- ownCloud config: read-write access
- ownCloud data: read-write access
- ownCloud database: complete database server ownage, including all users and databases.

Initially one might thing think Snappy's database access means it's less secure, but one needs to remember that the snap owns its own database server, and it only contains the ownCloud database. Complete ownage of that server is no greater compromise than that of the traditional install's.

Long story short, whether traditional or Snappy, a compromise of ownCloud only results in a compromise of ownCloud.

**Comment by oparoz** [comment]
Thanks for that @kyrofa .

It will take a bit of getting used-to and might require some complex vhosts for some webapps, but luckily it shouldn't be too much of a problem with ownCloud as the `.htaccess` stuff can be moved to the vhost.

**Comment by jospoortvliet** [comment]
So @oparoz has created a new PR which is... more likely to be useful ;-)
